compiler/src/dotty/tools/dotc/typer/Applications.scala

```diff
@@ -1427,7 +1427,17 @@ trait Applications extends Compatibility { self: Typer with Dynamic =>
         narrowByTypes(alts, args, resultType)
 
       case pt =>
-        alts filter (normalizedCompatible(_, pt))
+        val noSam = alts filter (normalizedCompatible(_, pt))
+        if (noSam.isEmpty) {
+          pt match {
+            case SAMType(mtp) =>
+              val sam = narrowByTypes(alts, mtp.paramInfos, mtp.resultType)
+              if (sam.nonEmpty && !pt.classSymbol.hasAnnotation(defn.FunctionalInterfaceAnnot))
+                ctx.warning(ex"$pt does not have the @FunctionalInterface annotation.", ctx.tree.pos)
```

Comment:
I still don't think this is worth adding a warning for. Even if it is, there should be a simple way to turn off this warning.

Reply:
The simple way is to do explicit eta-expension, no?
```scala
new ObjectOutputStream(println) // warning: are you sure this what you want to do?
new ObjectOutputStream(() => println()) // Ok
new ObjectOutputStream(println _)  // Ok
```